### PR TITLE
fix: remove redundant spaces when auto-fixing `prefer-template` rule

### PR DESCRIPTION
**Pull request by galabra**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Following issue [https://github.com/eslint/eslint/issues/8520](https://github.com/eslint/eslint/issues/8520) (one of multiple issues that describe of the same topic), I added a logic to ignore redundant spaces when auto-fixing the [`prefer-template`](https://eslint.org/docs/latest/rules/prefer-template) rule.
The change affects pretty much all of the invalid case tests, hence it's self-explanatory.

#### Is there anything you'd like reviewers to focus on?

I wasn't sure how much of the code should be inline, as opposed to helper functions.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
@galabra thanks for the PR, but I'm not sure if this change is necessary. Doesn't [template-curly-spacing](https://eslint.org/docs/latest/rules/template-curly-spacing) solve the problem with spaces that remain after `prefer-template` fixes the code?

**Comment by galabra** [comment]
@mdjermanovic thank you for the quick comment!

Let me start by admitting you _are_ mathematically correct. When running `eslint --fix` it'll eventually get fixed.
However, from a **Developer User Experience** point of view, I believe there's no reason to keep this behavior of `prefer-template`. It serves absolutely no purpose, looks buggy and honestly just makes this rule seem "half baked".

You can tell by the amount of issues that discussed this over the years (https://github.com/microsoft/vscode-eslint/issues/189, https://github.com/microsoft/vscode-eslint/issues/1001, https://github.com/prettier/eslint-plugin-prettier/issues/153, https://github.com/eslint/eslint/issues/10025) that this is a questionable behavior, to say the least.

File: lib/rules/prefer-template.js [hunk 1]
```diff
@@ -192,32 +195,35 @@ module.exports = {
...
                 /*
                  * Otherwise, these nodes should not be combined into a template curly, since there is nowhere to put
                  * the text between them.
                  */
-                return `${getTemplateLiteral(currentNode.left, textBeforeNode, null)}${textBeforePlus}+${textAfterPlus}${getTemplateLiteral(currentNode.right, textAfterNode, null)}`;
+                return `${getTemplateLiteral(currentNode.left, cleanTextBeforeNode, null)}${textBeforePlus}+${textAfterPlus}${getTemplateLiteral(currentNode.right, cleanTextAfterNode, null)}`;
             }
 
-            return `\`\${${textBeforeNode || ""}${sourceCode.getText(currentNode)}${textAfterNode || ""}}\``;
+            return `\`\${${cleanTextBeforeNode}${sourceCode.getText(currentNode)}${cleanTextAfterNode}}\``;
```
Comment (mdjermanovic) [inline lib/rules/prefer-template.js:226]: Would it work the same if we trim spaces only at this point?

```diff
-return `\`\${${textBeforeNode || ""}${sourceCode.getText(currentNode)}${textAfterNode || ""}}\``;
+return `\`\${${(textBeforeNode || "").replace(/^ +/u, "")}${sourceCode.getText(currentNode)}${(textAfterNode || "").replace(/ +$/u, "")}}\``;
```

Comment (galabra) [reply lib/rules/prefer-template.js:226]: The `cleanTextAroundPlus` is required for cases with the `+` operator. For example:
```
var foo = 'bar' + baz;
// turns into
var foo = `bar${ baz}`;
// but it should be
var foo = `bar${baz}`;
```

But you're definitely right about line 223! I reverted this change

**Comment by galabra** [comment]
@mdjermanovic kindly reminding about this 🙏
Much appreciated!

**Comment by mdjermanovic** [comment]
In some cases, the formatting looks worse after this change.

Example:

```js
const foo = bar(
    a +
    "b" +
    c
);
```

after the auto-fix:

```js
const foo = bar(
    `${a
}b${
    c}`
);

```

Example: 

```js
const foo = bar(
    a
    + "b"
    + c
);


```

after the auto-fix:

```js
const foo = bar(
    `${a
}b${
c}`
);


```

**Comment by nzakas** [comment]
I'm not sure this is worth fixing. Part of the reason that we haven't made this change already is precisely as @mdjermanovic has mentioned: it creates other spacing issues. Trading one set of spacing issues for another set doesn't make a lot of sense to me. Our rules are meant to be make fixes as narrowly as possible to minimize bugs and we explicitly try not to do any formatting inside of fixes because that formatting could very well be wrong based on how other rules are configured.

**Comment by mdjermanovic** [comment]
I agree with @nzakas, hence closing. It seems best to leave formatting to other rules. Anyway, thanks for the attempt!

**Comment by abrindam** [comment]
I know this has been closed for a couple months, but I wanted to provide feedback that I disagree with this decision.

I completely understand the concern around the more complicated formatting cases and not wanting to implement formatting logic inside a rule that is not aimed at formatting concerns. However, I feel that the wholesale rejection of this idea is suffering from throwing out the baby with the bath water.

I am fairly certain a significant percentage of the times this rule comes into play, it is one of these trivial cases:
```
const foo = "bar" + qux
const foo =  qux + "bar
const foo = "bar" + qux + "baz"
```

In all these cases, the output of the auto-fix will contain multiple leading and/or trailing spaces inside the curly brackets. In my opinion, this resulting formatting is objectively odd-looking, and is not a formatting decision that any human would intentionally make.

The most likely result of this is that an ESLint user, like myself, will assume this is a bug, and will lose faith in the tool - something that would be very regrettable. It might also cause them to either disable this rule and/or disable auto-fix-on-save, since the generated code is ugly.

I understand the counter-argument that all you have to do is enable `template-curly-spacing` and/or `prettier` and this problem goes away. Practically speaking, however, this may not be possible. In my case we are first rolling out linting rules, then formatting rules, so I am stuck with this issue in the interim. From a theoretical perspective, I would also argue that the "correctness" of a rule should not be impacted by what other rules are enabled (again, I argue that the generated code is so odd looking as to be considered seemingly incorrect, even if it does compile).

As I am sympathetic to the more complicated formatting concerns, would you consider a modified version of this PR which only fixes this for the most trivial cases I listed above? I think that once a user introduces newlines, etc in the middle of string concatenation, they have no reasonable expectation of the code formatting, but in these trivial cases the current behavior is likely to be out of line with anyone's expectations.

Thanks for considering!

**Pull request closed without merging**